lms/views/api/canvas/authorize.py

```diff
@@ -21,19 +21,31 @@
 @view_defaults(request_method="GET", route_name="canvas_oauth_callback")
 class CanvasAPIAuthorizeViews:
 
-    #: The Canvas API scopes that we request.
-    scopes = (
+    # The Canvas API scopes that we need for our Canvas Files feature.
+    files_scopes = (
         "url:GET|/api/v1/courses/:course_id/files",
         "url:GET|/api/v1/files/:id/public_url",
     )
 
+    # The Canvas API scopes that we need for our Sections feature.
+    sections_scopes = (
+        "url:GET|/api/v1/courses/:id",
+        "url:GET|/api/v1/courses/:course_id/sections",
+        "url:GET|/api/v1/courses/:course_id/users/:id",
+    )
+
     def __init__(self, request):
         self.request = request
 
     @view_config(permission="canvas_api", route_name="canvas_api.authorize")
     def authorize(self):
         ai_getter = self.request.find_service(name="ai_getter")
 
+        if ai_getter.canvas_sections_enabled():
+            scopes = self.files_scopes + self.sections_scopes
+        else:
+            scopes = self.files_scopes
```

Comment:
We're calling `ai_getter.canvas_sections_enabled()` here because we _can't_ call `LTILaunchResource.canvas_sections_enabled()`: `LTILaunchResource` is not the context resource for this view. We can't extract `LTILaunchResource.canvas_sections_enabled()` out of that class either: it depends in `is_canvas`, which depends on checking certain LTI launch params to decide whether we're in Canvas or not, and those launch params are only present during the original launch request.

Here, in the authorization popup window, we're not "in Canvas" (where "in Canvas" is defined as the request has certain launch params that identify it as from Canvas). We're in a standalone popup window.

Fortunately this popup window would never be opened in a non-Canvas LMS anyway (it's the Canvas API authorization popup) so it's not necessary to check whether we're in Canvas here.

Reply:
If we had tried to prematurely generalize this we'd be completely fucked right now